src/printer.cpp

```diff
@@ -0,0 +1,314 @@
+#include <algorithm>
+
+#include "printer.h"
+
+namespace CSymPy {
+
+StrPrinter::StrPrinter(bool ordered) : BaseVisitor(this), order_(ordered) {
+
+}
+void StrPrinter::bvisit(const Basic &x) {
+    std::cout << "Basic" <<std::endl;
+}
+
+void StrPrinter::bvisit(const Symbol &x) {
+    str_ = x.get_name();
+}
+
+void StrPrinter::bvisit(const Integer &x) {
+    std::ostringstream s;
+    s << x.i;
+    str_ = s.str();
+}
+
+void StrPrinter::bvisit(const Rational &x) {
+    std::ostringstream s;
+    s << x.i;
+    str_ = s.str();
+}
+
+void StrPrinter::bvisit(const Complex &x) {
+    std::ostringstream s;
+    if (x.real_ != 0) {
+        s << x.real_;
+        // Since Complex is in canonical form, imaginary_ is not 0.
+        if (sgn(x.imaginary_) == 1) {
+            s << " + ";
+        } else {
+            s << " - ";
+        }
+        // If imaginary_ is not 1 or -1, print the absolute value
+        if (x.imaginary_ != sgn(x.imaginary_)) {
+            s << abs(x.imaginary_);
+            s << "*I";
+        } else {
+            s << "I";
+        }
+    } else {
+        if (x.imaginary_ != sgn(x.imaginary_)) {
+            s << x.imaginary_;
+            s << "*I";
+        } else {
+            if (sgn(x.imaginary_) == 1) {
+                s << "I";
+            } else {
+                s << "-I";
+            }
+        }
+    }
+    str_ = s.str();
+}
+
+template<class T>
+void getKeysValues(vec_basic &keys, vec_basic &values, T &dict, bool order) {
+
+    if(order) {
+        for(auto &p: dict) {
+            keys.push_back(p.first);
+        }
+        std::sort(keys.begin(), keys.end(), RCPBasicKeyLessCmp());
+        for(auto &p: keys) {
+            values.push_back(dict.find(p)->second);
+        }
```

Comment:
By old order, you mean the random order? I wouldn't keep it. If somebody
needs a fast printer, they can write a specialized printer.

Sent from my mobile phone.
On Apr 17, 2015 10:35 PM, "Isuru Fernando" <notifications@github.com> wrote:

> In src/printer.cpp
> <https://github.com/sympy/csympy/pull/423#discussion_r28641279>:
>
> > +        }
> > +    }
> > +    str_ = s.str();
> > +}
> > +
> > +template<class T>
> > +void getKeysValues(vec_basic &keys, vec_basic &values, T &dict, bool order) {
> > +
> > +    if(order) {
> > +        for(auto &p: dict) {
> > +            keys.push_back(p.first);
> > +        }
> > +        std::sort(keys.begin(), keys.end(), RCPBasicKeyLessCmp());
> > +        for(auto &p: keys) {
> > +            values.push_back(dict.find(p)->second);
> > +        }
>
> Yeah, I wanted to find a common way for both orderings. I couldn't find a
> better way to make it work with both orderings.
> Btw, do we need to keep the earlier ordering. If it's just the new order,
> then your suggestion can be used.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/csympy/pull/423/files#r28641279>.
>